File: java/src/main/java/com/authdog/resources/UsersResource.java
package com.authdog.resources;

import com.authdog.AuthdogClient;
import com.authdog.exceptions.ApiException;
import com.authdog.exceptions.AuthenticationException;
import com.authdog.types.EnvUsersResponse;
import com.fasterxml.jackson.databind.JsonNode;
import java.util.Map;

/**
 * Directory user operations.
 */
public final class UsersResource {
    /**
     * Authdog client.
     */
    private final AuthdogClient client;

    /**
     * Create a users helper.
     * @param clientParam Authdog client
     */
    public UsersResource(final AuthdogClient clientParam) {
        this.client = clientParam;
    }

    /**
     * List users in an environment.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @return users envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public EnvUsersResponse list(final String tenantId,
                                 final String environmentId)
            throws AuthenticationException, ApiException {
        return list(tenantId, environmentId, null, null, null);
    }

    /**
     * List users in an environment with optional filters.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param offset page offset
     * @param limit page size
     * @param searchQuery search query
     * @return users envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public EnvUsersResponse list(final String tenantId,
                                 final String environmentId,
                                 final Integer offset,
                                 final Integer limit,
                                 final String searchQuery)
            throws AuthenticationException, ApiException {
        final Map<String, String> params = QueryParams.create();
        QueryParams.put(params, "offset", offset);
        QueryParams.put(params, "limit", limit);
        QueryParams.put(params, "searchQuery", searchQuery);
        return client.request("GET",
                "/v1/tenants/" + tenantId + "/environments/"
                        + environmentId + "/users",
                null, QueryParams.orNull(params),
                EnvUsersResponse.class);
    }

    /**
     * Create a user.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param body request body
     * @return created user envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode create(final String tenantId,
                           final String environmentId,
                           final Object body)
            throws AuthenticationException, ApiException {
        return client.request("POST",
                "/v1/tenants/" + tenantId + "/environments/"
                        + environmentId + "/users",
                body, null, JsonNode.class);
    }

    /**
     * Search users in an environment.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param q search query
     * @return users envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public EnvUsersResponse search(final String tenantId,
                                   final String environmentId,
                                   final String q)
            throws AuthenticationException, ApiException {
        return search(tenantId, environmentId, q, null, null);
    }

    /**
     * Search users in an environment with paging.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param q search query
     * @param offset page offset
     * @param limit page size
     * @return users envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public EnvUsersResponse search(final String tenantId,
                                   final String environmentId,
                                   final String q,
                                   final Integer offset,
                                   final Integer limit)
            throws AuthenticationException, ApiException {
        final Map<String, String> params = QueryParams.create();
        QueryParams.put(params, "q", q);
        QueryParams.put(params, "offset", offset);
        QueryParams.put(params, "limit", limit);
        return client.request("GET",
                "/v1/tenants/" + tenantId + "/environments/"
                        + environmentId + "/users/search",
                null, QueryParams.orNull(params),
                EnvUsersResponse.class);
    }

    /**
     * Count users in an environment.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @return count envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode count(final String tenantId,
                          final String environmentId)
            throws AuthenticationException, ApiException {
        return client.request("GET",
                "/v1/tenants/" + tenantId + "/environments/"
                        + environmentId + "/users/count",
                null, null, JsonNode.class);
    }

    /**
     * Get a user.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param userId user ID
     * @return user envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode get(final String tenantId,
                        final String environmentId,
                        final String userId)
            throws AuthenticationException, ApiException {
        return client.request("GET",
                "/v1/tenants/" + tenantId + "/environments/"
                        + environmentId + "/users/" + userId,
                null, null, JsonNode.class);
    }

    /**
     * Update a user's profile.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param userId user ID
     * @param body request body
     * @return update envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode update(final String tenantId,
                           final String environmentId,
                           final String userId, final Object body)
            throws AuthenticationException, ApiException {
        return client.request("PUT",
                "/v1/tenants/" + tenantId + "/environments/"
                        + environmentId + "/users/" + userId,
                body, null, JsonNode.class);
    }

    /**
     * Delete a user from the environment.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param userId user ID
     * @return delete envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode delete(final String tenantId,
                           final String environmentId,
                           final String userId)
            throws AuthenticationException, ApiException {
        return client.request("DELETE",
                "/v1/tenants/" + tenantId + "/environments/"
                        + environmentId + "/users/" + userId,
                null, null, JsonNode.class);
    }

    /**
     * Enable or disable a user.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param userId user ID
     * @param body request body
     * @return update envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode setActive(final String tenantId,
                              final String environmentId,
                              final String userId, final Object body)
            throws AuthenticationException, ApiException {
        return client.request("PATCH",
                "/v1/tenants/" + tenantId + "/environments/"
                        + environmentId + "/users/" + userId
                        + "/active",
                body, null, JsonNode.class);
    }

    /**
     * List groups a user belongs to.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param userId user ID
     * @return groups envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode listGroups(final String tenantId,
                               final String environmentId,
                               final String userId)
            throws AuthenticationException, ApiException {
        return client.request("GET",
                "/v1/tenants/" + tenantId + "/environments/"
                        + environmentId + "/users/" + userId
                        + "/groups",
                null, null, JsonNode.class);
    }
}

File: csharp/Resources/UsersResource.cs
using System.Net.Http;
using System.Threading.Tasks;
using Authdog.Types;
using Newtonsoft.Json.Linq;

namespace Authdog
{
    /// <summary>
    /// Directory user Wave 1 operations
    /// </summary>
    public class UsersResource
    {
        private readonly AuthdogClient _client;

        public UsersResource(AuthdogClient client)
        {
            _client = client;
        }

        public Task<EnvUsersResponse> ListAsync(
            string tenantId,
            string environmentId,
            int? offset = null,
            int? limit = null,
            string? searchQuery = null) =>
            _client.RequestAsync<EnvUsersResponse>(
                HttpMethod.Get,
                $"/v1/tenants/{tenantId}/environments/{environmentId}/users",
                query: AuthdogClient.Params(
                    ("offset", offset),
                    ("limit", limit),
                    ("searchQuery", searchQuery)));

        public EnvUsersResponse List(
            string tenantId,
            string environmentId,
            int? offset = null,
            int? limit = null,
            string? searchQuery = null) =>
            ListAsync(tenantId, environmentId, offset, limit, searchQuery).GetAwaiter().GetResult();

        public Task<JObject> CreateAsync(string tenantId, string environmentId, object body) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Post,
                $"/v1/tenants/{tenantId}/environments/{environmentId}/users",
                body);

        public JObject Create(string tenantId, string environmentId, object body) =>
            CreateAsync(tenantId, environmentId, body).GetAwaiter().GetResult();

        public Task<EnvUsersResponse> SearchAsync(
            string tenantId,
            string environmentId,
            string? q = null,
            int? offset = null,
            int? limit = null) =>
            _client.RequestAsync<EnvUsersResponse>(
                HttpMethod.Get,
                $"/v1/tenants/{tenantId}/environments/{environmentId}/users/search",
                query: AuthdogClient.Params(("q", q), ("offset", offset), ("limit", limit)));

        public EnvUsersResponse Search(
            string tenantId,
            string environmentId,
            string? q = null,
            int? offset = null,
            int? limit = null) =>
            SearchAsync(tenantId, environmentId, q, offset, limit).GetAwaiter().GetResult();

        public Task<JObject> CountAsync(string tenantId, string environmentId) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Get,
                $"/v1/tenants/{tenantId}/environments/{environmentId}/users/count");

        public JObject Count(string tenantId, string environmentId) =>
            CountAsync(tenantId, environmentId).GetAwaiter().GetResult();

        public Task<JObject> GetAsync(string tenantId, string environmentId, string userId) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Get,
                $"/v1/tenants/{tenantId}/environments/{environmentId}/users/{userId}");

        public JObject Get(string tenantId, string environmentId, string userId) =>
            GetAsync(tenantId, environmentId, userId).GetAwaiter().GetResult();

        public Task<JObject> UpdateAsync(string tenantId, string environmentId, string userId, object body) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Put,
                $"/v1/tenants/{tenantId}/environments/{environmentId}/users/{userId}",
                body);

        public JObject Update(string tenantId, string environmentId, string userId, object body) =>
            UpdateAsync(tenantId, environmentId, userId, body).GetAwaiter().GetResult();

        public Task<JObject> DeleteAsync(string tenantId, string environmentId, string userId) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Delete,
                $"/v1/tenants/{tenantId}/environments/{environmentId}/users/{userId}");

        public JObject Delete(string tenantId, string environmentId, string userId) =>
            DeleteAsync(tenantId, environmentId, userId).GetAwaiter().GetResult();

        public Task<JObject> SetActiveAsync(string tenantId, string environmentId, string userId, object body) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Patch,
                $"/v1/tenants/{tenantId}/environments/{environmentId}/users/{userId}/active",
                body);

        public JObject SetActive(string tenantId, string environmentId, string userId, object body) =>
            SetActiveAsync(tenantId, environmentId, userId, body).GetAwaiter().GetResult();

        public Task<JObject> ListGroupsAsync(string tenantId, string environmentId, string userId) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Get,
                $"/v1/tenants/{tenantId}/environments/{environmentId}/users/{userId}/groups");

        public JObject ListGroups(string tenantId, string environmentId, string userId) =>
            ListGroupsAsync(tenantId, environmentId, userId).GetAwaiter().GetResult();

        public Task<JObject> RevokeSessionAsync(string environmentId, string sessionId) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Delete,
                $"/v1/environments/{environmentId}/sessions/{sessionId}");

        public JObject RevokeSession(string environmentId, string sessionId) =>
            RevokeSessionAsync(environmentId, sessionId).GetAwaiter().GetResult();

        public Task<JObject> TotpStatusAsync(string tenantId, string environmentId) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Get,
                $"{AuthdogClient.Env(tenantId, environmentId)}/me/mfa/totp");

        public JObject TotpStatus(string tenantId, string environmentId) =>
            TotpStatusAsync(tenantId, environmentId).GetAwaiter().GetResult();

        public Task<JObject> BulkDeleteAsync(string tenantId, string environmentId, object body) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Post,
                $"{AuthdogClient.Env(tenantId, environmentId)}/users/bulk/delete",
                body);

        public JObject BulkDelete(string tenantId, string environmentId, object body) =>
            BulkDeleteAsync(tenantId, environmentId, body).GetAwaiter().GetResult();

        public Task<JObject> BulkSetActiveAsync(string tenantId, string environmentId, object body) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Post,
                $"{AuthdogClient.Env(tenantId, environmentId)}/users/bulk/set-active",
                body);

        public JObject BulkSetActive(string tenantId, string environmentId, object body) =>
            BulkSetActiveAsync(tenantId, environmentId, body).GetAwaiter().GetResult();

        public Task<JObject> ImportUsersAsync(string tenantId, string environmentId, object body) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Post,
                $"{AuthdogClient.Env(tenantId, environmentId)}/users/import",
                body);

        public JObject ImportUsers(string tenantId, string environmentId, object body) =>
            ImportUsersAsync(tenantId, environmentId, body).GetAwaiter().GetResult();

        public Task<JObject> DisableMfaAsync(string tenantId, string environmentId, string userId) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Delete,
                $"{AuthdogClient.Env(tenantId, environmentId)}/users/{userId}/mfa");

        public JObject DisableMfa(string tenantId, string environmentId, string userId) =>
            DisableMfaAsync(tenantId, environmentId, userId).GetAwaiter().GetResult();

        public Task<JObject> ListSessionsAsync(string tenantId, string environmentId, string userId) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Get,
                $"{AuthdogClient.Env(tenantId, environmentId)}/users/{userId}/sessions");

        public JObject ListSessions(string tenantId, string environmentId, string userId) =>
            ListSessionsAsync(tenantId, environmentId, userId).GetAwaiter().GetResult();
    }
}
